bitrig integration

dhuseby opened the pull request:
This patch adds the necessary pieces to support rust on Bitrig https://bitrig.org

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @pcwalton (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way Github handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [CONTRIBUTING.md](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

dhuseby commented:
Fixed the whitespace errors.

alexcrichton commented on mk/platform.mk:
```diff
@@ -144,14 +144,14 @@ FIND_COMPILER = $(word 1,$(1:ccache=))
+  RUSTC_CROSS_FLAGS_$(1)=-C linker=$$(call FIND_COMPILER,$$(CC_$(1))) \
+      -C ar=$$(call FIND_COMPILER,$$(AR_$(1))) $(RUSTC_CROSS_FLAGS_$(1))
+
+  RUSTC_FLAGS_$(1)=$$(RUSTC_CROSS_FLAGS_$(1)) $(RUSTC_FLAGS_$(1))
```
Can this keep the previous indentation?

alexcrichton commented on mk/platform.mk:
```diff
@@ -192,7 +200,7 @@ define CFG_MAKE_TOOLCHAIN
   # FIXME: We should be able to use the LLVM assembler
   CFG_ASSEMBLE_$(1)=$$(CC_$(1)) $$(CFG_GCCISH_CFLAGS_$(1)) \
-		    $$(CFG_DEPEND_FLAGS) $$(2) -c -o $$(1)
+        $$(CFG_DEPEND_FLAGS) $$(2) -c -o $$(1)
```
(same here)

alexcrichton commented on src/libstd/env.rs:
```diff
@@ -563,6 +563,70 @@ pub mod consts {
+}
+
+/// Constants associated with the current target
+#[cfg(target_os = "openbsd")]
```
Hasn't openbsd already been added? (maybe a rebase is needed?)

dhuseby replied:
So Bitrig is a fork of OpenBSD and was broken by the recent env patch.  This fixes that.  I just saw that there is a PR from the OpenBSD maintainer that also fixes this that hasn't landed yet: https://github.com/rust-lang/rust/pull/21964/files

dhuseby replied:
Oh wait, no it doesn't.

semarie replied:
I am sorry to don't have noted this pull request,and the correction you have include in it for OpenBSD, before I push the correction in #21964.

alexcrichton commented on src/libstd/sys/common/stack.rs:
```diff
@@ -189,11 +189,18 @@ pub unsafe fn record_sp_limit(limit: uint) {
+    #[inline(always)]
+    unsafe fn target_record_sp_limit(_: uint) {
+        // segmented stacks are disabled
+    }
```
There's a catch-all impl down below, could the `cfg` be added to that?

alexcrichton commented on src/libstd/sys/common/stack.rs:
```diff
@@ -276,12 +283,18 @@ pub unsafe fn get_sp_limit() -> uint {
+    #[inline(always)]
+    unsafe fn target_get_sp_limit() -> uint {
+      return 2048;
+    }
```
Could this be added to the `#[cfg]` down below?

alexcrichton commented:
r? @brson

dhuseby commented:
Let me fix the other issues and push again.

dhuseby commented:
I rebased to master too.

dhuseby commented:
I did.  In ad45164.

brson commented:
@bors: r+ fc147f1

semarie commented on src/liblibc/lib.rs:
```diff
@@ -1523,19 +1543,15 @@ pub mod types {
+                    pub st_atim: timespec,
+                    pub st_mtim: timespec,
+                    pub st_ctim: timespec,
                     pub st_size: off_t,
```
I am not completely agreed with the conversion form `time_t + c_long` to `timespec`.
Under bitrig and openbsd, the timespec definition is under conditionnal definition:
`#if __POSIX_VISIBLE >= 200809 || __BSD_VISIBLE`. Here, the mod is `posix01`.

Note: the C size of `timespec` is the same as `time_t + c_long`, so it don't break code, only the ABI.

see https://github.com/bitrig/bitrig/blob/master/sys/sys/stat.h#L53

dhuseby replied:
good point. 2001 < 2008

semarie commented on src/liblibc/lib.rs:
```diff
@@ -1523,19 +1543,15 @@ pub mod types {
                     pub st_gen: uint32_t,
-                    pub st_birthtime: time_t,
-                    pub st_birthtime_nsec: c_long,
+                    pub __st_birthtim: timespec,
```
Same comment here

semarie commented on src/librustc_back/x86.rs:
```diff
@@ -45,10 +45,7 @@ pub fn get_target_strs(target_triple: String, target_os: abi::Os)
-          abi::OsDragonfly => {
+          abi::OsFreebsd | abi::OsDragonfly | abi::OsBitrig => {
             "e-p:32:32-f64:32:64-i64:32:64-f80:32:32-n8:16:32".to_string()
           }
```
If you collapse `abi::OsFreebsd` + `abi::OsDragonfly` with `abi::OsBitrig`, could you include also `abi::OsOpenbsd` which is just under ? Thanks.

semarie commented on src/librustc_back/x86_64.rs:
```diff
@@ -47,12 +47,7 @@ pub fn get_target_strs(target_triple: String, target_os: abi::Os) -> target_strs
+          abi::OsFreebsd | abi::OsDragonfly | abi::OsBitrig => {
             "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-\
                 f32:32:32-f64:64:64-v64:64:64-v128:128:128-a:0:64-\
                 s0:64:64-f80:128:128-n8:16:32:64-S128".to_string()
```
Same here, about collapsing `abi::OsFreebsd | abi::OsDragonfly | abi::OsBitrig`.

semarie commented on src/libstd/sys/unix/c.rs:
```diff
@@ -336,3 +345,41 @@ mod signal {
 }
+
+#[cfg(target_os = "bitrig")]
+mod signal {
```
Oh, I note you diverge from openbsd because the openbsd is wrong for `struct siginfo`.
Could you include target_os="openbsd" here (and remove of it from previous block ?)

If not possible, not a problem, I will correct it after your PR will merged.

semarie commented on src/libstd/sys/unix/stack_overflow.rs:
```diff
@@ -219,6 +222,9 @@ mod imp {
 
+        #[cfg(target_os = "bitrig")]
+        pub const SIGSTKSZ: libc::size_t = 40960;
+
```
maybe collapse bitrig and openbsd in the same block ? but it is just cosmetic...

dhuseby commented:
@Manishearth good catch!  oops!

dhuseby commented:
Alright, all review changes are made and I rebased to master.

Manishearth commented:
The rollup should land in a few minutes; it has conflicts with your PR : try rebasing over https://github.com/Manishearth/rust/compare/rollup (https://github.com/rust-lang/rust/commit/d3732a12e896ab98aa27eaffab99a78bbaf837e4)

dhuseby commented:
@Manishearth I just rebased to d3732a1.

Manishearth commented:
Great, looks mergeable now.

semarie commented on src/liblibc/lib.rs:
```diff
@@ -4130,7 +4149,7 @@ pub mod consts {
             pub const _SC_TTY_NAME_MAX : c_int = 107;
-            pub const _SC_ATEXIT_MAX : c_int = 46;
+            pub const _SC_ATEXIT_MAX : c_int = 146;
             pub const _SC_XOPEN_CRYPT : c_int = 117;
```
the _SC_ATEXIT_MAX value is wrong for openbsd and for bitrig too.
please keep `46`.
see https://github.com/bitrig/bitrig/blob/master/include/unistd.h#L199

dhuseby replied:
Good catch again!  Thank you for taking the time to look at my patch so closely.  I must have added the '1' accidentally when I was going through the constants.

semarie commented on src/librustc_back/x86_64.rs:
```diff
@@ -47,17 +47,7 @@ pub fn get_target_strs(target_triple: String, target_os: abi::Os) -> target_strs
-          }
-          abi::OsOpenbsd => {
+          abi::OsFreebsd | abi::OsDragonfly | abi::OsBitrig | abi::Openbsd => {
             "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-\
```
`abi:Openbsd` will not been defined: it is `abi:OsOpenbsd`

dhuseby replied:
Yup...typo fixed in 8d76733

semarie commented on src/libstd/env.rs:
```diff
@@ -570,7 +602,7 @@ pub mod consts {
     /// A string describing the specific operating system in use: in this
-    /// case, `dragonfly`.
+    /// case, `openbsd`.
     pub const OS: &'static str = "openbsd";
```
yeah, thanks :)

dhuseby replied:
np :)

semarie commented on src/liblibc/lib.rs:
```diff
@@ -1502,8 +1521,9 @@ pub mod types {
                 use types::os::arch::posix88::{mode_t, off_t};
                 use types::os::arch::posix88::{uid_t};
+                use types::os::common::posix01::{timespec};
 
```
`timespec` is an unused import, and it is an error under `--cfg stage1` build.
`error: unused import, #[deny(unused_imports)] on by default`.

brson commented:
Needs rebase.

dhuseby commented:
rebased.

brson commented:
@bors: r+ bd70

bors commented:
:hourglass: Testing commit bd704f5 with merge 48e9eb6...

bors commented:
:broken_heart: Test failed - [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/3557)

dhuseby commented:
I don't get it.  I don't think I changed anything that touched the mac os build.

alexcrichton commented:
It looks like there was a compile error with liblibc, perhaps some imports got lost in the shuffling around?

dhuseby commented:
@alexcrichton yup, there was a merge error earlier that duplicated the macos block of constants.  i picked the wrong one.  i fixed the problem in 5513040 and rebased to master HEAD.

brson commented:
@bors: r+ 5513

bors commented:
:hourglass: Testing commit 5513040 with merge ab5d389...

bors commented:
:broken_heart: Test failed - [auto-win-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-32-nopt-t/builds/3268)

bors commented:
:broken_heart: Test failed - [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/3574)

dhuseby commented:
@bors the test failures are both erroneous. the android build was interrupted and the windows build failed to remove a file somewhere.  neither of which appear to be related to my work.

alexcrichton commented:
@bors: retry

bors commented:
:hourglass: Testing commit 5513040 with merge cdc6379...

bors commented:
:broken_heart: Test failed - [auto-win-64-opt](http://buildbot.rust-lang.org/builders/auto-win-64-opt/builds/2137)

alexcrichton commented:
@bors: retry

bors commented:
:hourglass: Testing commit 5513040 with merge 14ac4d0...

bors commented:
:broken_heart: Test failed - [auto-win-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-32-nopt-t/builds/3320)

alexcrichton commented:
@bors: retry

bors commented:
:hourglass: Testing commit 5513040 with merge 603553a...

bors commented:
:broken_heart: Test failed - [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/3674)

dhuseby commented:
@alexcrichton @brson again, the auto-mac-32-opt failures are unrelated to my work.  the first error is a panic in ReaderRng.fill_bytes:

```
thread '<unnamed>' panicked at 'ReaderRng.fill_bytes error: IoError { kind: EndOfFile, desc: "end of file", detail: None }', /Users/rustbuild/src/rust-buildbot/slave/auto-mac-32-opt/build/src/libstd/rand/reader.rs:70
```

and then there are some test failures:

```
<test>:1:62: 1:63 error: macro undefined: 'z!'
<test>:1 fn bogus() {macro_rules! z (() => (3+4));}fn inty() -> i32 { z!() }
                                                                      ^
error: aborting due to previous error
<test>:1:59: 1:60 error: macro undefined: 'z!'
<test>:1 mod foo {macro_rules! z (() => (3+4));}fn inty() -> i32 { z!() }
                                                                   ^
error: aborting due to previous error
```

I'm not sure what to do about these?  They are unrelated to my changes.  Should these prevent my changes from landing?

dhuseby commented:
@alexcrichton @brson should I rebase to get updated code that might fix the unrelated test failures?

alexcrichton commented:
@bors: retry

Oh this just looks like a spurious timeout failure

bors commented:
:hourglass: Testing commit 5513040 with merge 2f3f7d5...

bors commented:
:broken_heart: Test failed - [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/3721)

dhuseby commented:
@alexcrichton @brson again, I'm not sure why this test is failing:

```
thread '<unnamed>' panicked at 'ReaderRng.fill_bytes error: IoError { kind: EndOfFile, desc: "end of file", detail: None }', /Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-opt/build/src/libstd/rand/reader.rs:70
```

I'm certain that this is unrelated to my patch.

dhuseby commented:
would rebasing help here?

semarie commented:
@dhuseby the error you cite isn't the cause of the build failed. I think the error line come from `rand::reader::test::test_reader_rng_insufficient_bytes` test (which is ok).

The problem is a timeout in testing std crate (20 min without output).

I am not sure a rebasing would help. it seems some buildbots have some instabilities (timeout, file creation problem...).

alexcrichton commented:
@bors: retry

@semarie's analysis is spot on (this just needs a retry)

dhuseby commented:
@alexcrichton how many more times do we have to repeat this before we declare the testing system broken and land the patch anyway?

alexcrichton commented:
Sorry our buildbots aren't always the most reliable builders :(

brson commented:
@bors: retry

Apparently there have been a lot of timeouts lately, but unfortunately we can't land anything without a successful test run.

brson commented:
I don't see other mac timeouts from looking through the build history.

@alexcrichton says we're having a lot of problems with windows right now.

bors commented:
:hourglass: Testing commit 5513040 with merge 9efabd8...

bors commented:
:broken_heart: Test failed - [auto-win-64-opt](http://buildbot.rust-lang.org/builders/auto-win-64-opt/builds/2230)

alexcrichton commented:
@bors: retry

bors commented:
:hourglass: Testing commit 5513040 with merge 2b01a37...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/3727), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/3733), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/3722), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/3731), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/3704), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/3745), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/3749), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/3743), [auto-win-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-32-nopt-t/builds/3401), [auto-win-32-opt](http://buildbot.rust-lang.org/builders/auto-win-32-opt/builds/3394), [auto-win-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-64-nopt-t/builds/2225), [auto-win-64-opt](http://buildbot.rust-lang.org/builders/auto-win-64-opt/builds/2234)
